pkg/controller/resourcedistribution/resourcedistribution_controller.go

```diff
@@ -0,0 +1,368 @@
+				return
+			}
+
+			// 2. if resource doesn't exist, create resource;
+			newResource := makeResourceObject(distributor, namespace, resource, newResourceHashCode)
+			if getErr != nil && errors.IsNotFound(getErr) {
+				if createErr := r.Client.Create(context.TODO(), newResource); createErr != nil {
+					if errors.IsNotFound(createErr) {
+						errCh <- &UnexpectedError{
+							err:         fmt.Errorf("namespace not found"),
+							namespace:   namespace,
+							conditionID: NoUncreatedNSConditionID,
+						}
+					} else {
+						errCh <- &UnexpectedError{
+							err:         createErr,
+							namespace:   namespace,
+							conditionID: CreateConditionID,
+						}
+					}
+					klog.Errorf("Error occurred when creating resource in namespace %s, err： %v, name: %s", namespace, createErr, distributor.Name)
+					return
+				}
+				atomic.AddInt32(&succeeded, 1)
+				klog.V(3).Infof("ResourceDistribution(%s) created (%s/%s) in namespaces %s", distributor.Name, resourceKind, resourceName, namespace)
+				return
+			}
+
+			// 3. if resource exits
+			annotations := oldResource.GetAnnotations()
+			if annotations == nil || annotations[utils.SourceResourceDistributionOfResource] != distributor.Name {
+				// if conflict occurred
+				errCh <- &UnexpectedError{
+					err:         fmt.Errorf("conflict with existing resources because of the same namespace and name"),
+					namespace:   namespace,
+					conditionID: NoConflictConditionID,
+				}
+				klog.Errorf("Conflict with existing resource(%s/%s) in namespaces %s, name: %s", resourceKind, resourceName, namespace, distributor.Name)
+				return
+			} else if annotations[utils.ResourceHashCodeAnnotation] != newResourceHashCode {
+				//  else if the resource needs to update
+				if updateErr := r.Client.Update(context.TODO(), newResource); updateErr != nil {
+					errCh <- &UnexpectedError{
+						err:         updateErr,
+						namespace:   namespace,
+						conditionID: UpdateConditionID,
+					}
+					klog.Errorf("Error occurred when updating resource in namespace %s, err： %v, name: %s", namespace, updateErr, distributor.Name)
+					return
+				}
+				klog.V(3).Infof("ResourceDistribution(%s) updated (%s/%s) for namespaces %s", distributor.Name, resourceKind, resourceName, namespace)
+			}
+			atomic.AddInt32(&succeeded, 1)
+		}(namespaceName)
+	}
+
+	wait.Wait() // wait for all distributions done.
+	return succeeded, errCh
+}
+
+func (r *ReconcileResourceDistribution) cleanResource(distributor *appsv1alpha1.ResourceDistribution,
+	unmatchedNamespaces sets.String, resource runtime.Object) <-chan *UnexpectedError {
+
+	wait := sync.WaitGroup{}
+	wait.Add(len(unmatchedNamespaces))
+	errCh := make(chan *UnexpectedError, len(unmatchedNamespaces))
+	for namespaceName := range unmatchedNamespaces {
+		go func(namespace string) {
+			defer wait.Done()
+
+			resourceName := utils.ConvertToUnstructured(resource).GetName()
```

Comment:
move ConvertToUnstructured  out of the namespace iteration. The resource should be almost identical for different namespace